### ISSUE TITLE
Suggestion: give a complete example of how to use BatchSet

### ISSUE THREAD

**Issue by GwynethLlewelyn** [dgraph-io/badger#224]
Since 1.0 is already out, and there has been a change to the `BatchSet` function parameters (see #98), I would like to request that the documentation also includes an example on how to properly use `BatchSet`to load millions of keys, especially avoiding the memory issues previously addressed by #163.

(Note that I've been attempting to force the GC to run each time a batch is processed by a goroutine, but I still easily break all memory limits; my current solution is to create the database on a machine with 'infinite' memory and upload the generated database to a server with strict memory constraints...)

**Comment by manishrjain** [comment]
Hey @GwynethLlewelyn , v1.0 is not out yet. We're gearing up for it, current ETA end Sep.

If you're running into memory issues, you can switch table loading parameters to MemoryMap. This would not load the LSM tree into RAM, hence decreasing the memory usage. Note that this is useful if you're doing bulk loading into Badger. For read-heavy workload, I'd recommend sticking to the default option of loading LSM tree into RAM.

https://godoc.org/github.com/dgraph-io/badger/options#FileLoadingMode

We have an example for BatchSetAsync:
https://godoc.org/github.com/dgraph-io/badger#example-KV-BatchSetAsync

Let me know if these pointers help. Or, what you like us to do w.r.t. documentation.

**Comment by GwynethLlewelyn** [comment]
Thanks for the pointers, @manishrjain — the example for BatchSetAsync certainly helps, although I'm curious — you seem just to be loading batches of 100 elements at each time? This is actual code from Dgraph, or just an example? My current default was loading 100 _thousand_ entries on each batch... In any case, I also tried to load batches of just 100 entries, and I hit the memory/CPU limitas quickly as with 100,000 entries. However, the way I figured out how to properly set up each batch is considerably more complicated than the example you've given, and very likely wrongly implemented (even though it certainly works in environments with 'unlimited' memory/CPU :) ). I'm going to rewrite the code — and possibly using BatchSetAsync as well, why not? — and see if I can get the kind of improvement that your own code is apparently able to get. In any case, thanks so much for looking into my question!

**Comment by manishrjain** [comment]
100K entries is definitely too many for a batch. Go GC doesn't respond well when you allocate and deallocate a lot of memory together. Whatever you deallocate, it would want to hold onto this memory and release it slowly to the OS -- while you're still allocating new big chunks of memory. This is what's probably causing your OOM issue.

You can do batches of 1000, and use BatchSetAsync. The good thing about using the async version is that your writes aren't blocked, and the actual writing can happen in another goroutine (within Badger); so you can push more throughput that way.

**Comment by GwynethLlewelyn** [comment]
I'm sorry for posting late — the issue is closed and may remain closed! — but I think it's nevertheless useful to relate my own experience (positive, in this case!) handling those 9 million keys I've got.
_Basically my problem was that I was reinventing the wheel by creating 'my' own version of `BatchSetAsync`_!!!
Once I figured out how to do the same using the example provided in the documentation... not only things became _way_ more simpler, but actual import time was reduced from 3 minutes to about 1 minute and 38s — same dataset, same parametres, just using`BatchSetAsync`instead of my own cooked-up version of it (which used `BatchSet` in a separate goroutine). Also, it seems to be using less memory as well.
Note that I'm _still_ using batches of 100,000 elements each time — that's the only way to get 9 million entries into the KV database in an acceptable interval of time 👍
